File: frontend/lib/domain/entities/dish.dart
// 料理関連のモデル

/// レシピ詳細
class RecipeDetails {
  final int? prepTime;
  final int? cookTime;
  final int? servings;
  final List<String> steps;
  final String? tips;
  final String? variations;

  RecipeDetails({
    this.prepTime,
    this.cookTime,
    this.servings,
    this.steps = const [],
    this.tips,
    this.variations,
  });

  factory RecipeDetails.fromJson(Map<String, dynamic> json) {
    return RecipeDetails(
      prepTime: json['prep_time'],
      cookTime: json['cook_time'],
      servings: json['servings'],
      steps: (json['steps'] as List?)?.map((e) => e.toString()).toList() ?? [],
      tips: json['tips'],
      variations: json['variations'],
    );
  }
}

/// 料理の材料
class DishIngredient {
  final int foodId;
  final String? foodName;
  final double amount;
  final String displayAmount;  // 表示用の量（例: "1", "1/2"）
  final String unit;           // 単位（例: "本", "個", "g"）
  final String cookingMethod;

  DishIngredient({
    required this.foodId,
    this.foodName,
    required this.amount,
    this.displayAmount = '',
    this.unit = 'g',
    this.cookingMethod = '生',
  });

  factory DishIngredient.fromJson(Map<String, dynamic> json) {
    return DishIngredient(
      foodId: json['food_id'],
      foodName: json['food_name'],
      amount: (json['amount'] ?? 0).toDouble(),
      displayAmount: json['display_amount'] ?? '',
      unit: json['unit'] ?? 'g',
      cookingMethod: json['cooking_method'] ?? '生',
    );
  }

  /// 表示用の量（単位付き + グラム表示）
  String get amountDisplay {
    // 単位がgやkg、mlの場合は重量のみ
    if (unit == 'g' || unit == 'kg' || unit == 'ml' || unit == 'L') {
      if (amount >= 1000) {
        return '${(amount / 1000).toStringAsFixed(1)}kg';
      }
      return '${amount.toStringAsFixed(0)}$unit';
    }
    // それ以外は「1本 (150g)」のような形式
    final gramDisplay = amount >= 1000
        ? '${(amount / 1000).toStringAsFixed(1)}kg'
        : '${amount.toStringAsFixed(0)}g';
    return '$displayAmount$unit ($gramDisplay)';
  }
}

/// 料理データモデル
class Dish {
  final int id;
  final String name;
  final String category;
  final List<String> mealTypes;
  final double servingSize;
  final String? description;
  final String? instructions;
  final List<DishIngredient> ingredients;
  final int storageDays;
  final int minServings;
  final int maxServings;
  // 基本栄養素
  final double calories;
  final double protein;
  final double fat;
  final double carbohydrate;
  final double fiber;
  // ミネラル
  final double sodium;
  final double potassium;
  final double calcium;
  final double magnesium;
  final double iron;
  final double zinc;
  // ビタミン
  final double vitaminA;
  final double vitaminD;
  final double vitaminE;
  final double vitaminK;
  final double vitaminB1;
  final double vitaminB2;
  final double vitaminB6;
  final double vitaminB12;
  final double niacin;
  final double pantothenicAcid;
  final double biotin;
  final double folate;
  final double vitaminC;
  final RecipeDetails? recipeDetails;

  Dish({
    required this.id,
    required this.name,
    required this.category,
    required this.mealTypes,
    this.servingSize = 1.0,
    this.description,
    this.instructions,
    this.ingredients = const [],
    this.storageDays = 1,
    this.minServings = 1,
    this.maxServings = 4,
    this.calories = 0,
    this.protein = 0,
    this.fat = 0,
    this.carbohydrate = 0,
    this.fiber = 0,
    this.sodium = 0,
    this.potassium = 0,
    this.calcium = 0,
    this.magnesium = 0,
    this.iron = 0,
    this.zinc = 0,
    this.vitaminA = 0,
    this.vitaminD = 0,
    this.vitaminE = 0,
    this.vitaminK = 0,
    this.vitaminB1 = 0,
    this.vitaminB2 = 0,
    this.vitaminB6 = 0,
    this.vitaminB12 = 0,
    this.niacin = 0,
    this.pantothenicAcid = 0,
    this.biotin = 0,
    this.folate = 0,
    this.vitaminC = 0,
    this.recipeDetails,
  });

  factory Dish.fromJson(Map<String, dynamic> json) {
    return Dish(
      id: json['id'],
      name: json['name'],
      category: json['category'],
      mealTypes: (json['meal_types'] as List?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      servingSize: (json['serving_size'] ?? 1.0).toDouble(),
      description: json['description'],
      instructions: json['instructions'],
      ingredients: (json['ingredients'] as List?)
              ?.map((e) => DishIngredient.fromJson(e))
              .toList() ??
          [],
      storageDays: json['storage_days'] ?? 1,
      minServings: json['min_servings'] ?? 1,
      maxServings: json['max_servings'] ?? 4,
      calories: (json['calories'] ?? 0).toDouble(),
      protein: (json['protein'] ?? 0).toDouble(),
      fat: (json['fat'] ?? 0).toDouble(),
      carbohydrate: (json['carbohydrate'] ?? 0).toDouble(),
      fiber: (json['fiber'] ?? 0).toDouble(),
      sodium: (json['sodium'] ?? 0).toDouble(),
      potassium: (json['potassium'] ?? 0).toDouble(),
      calcium: (json['calcium'] ?? 0).toDouble(),
      magnesium: (json['magnesium'] ?? 0).toDouble(),
      iron: (json['iron'] ?? 0).toDouble(),
      zinc: (json['zinc'] ?? 0).toDouble(),
      vitaminA: (json['vitamin_a'] ?? 0).toDouble(),
      vitaminD: (json['vitamin_d'] ?? 0).toDouble(),
      vitaminE: (json['vitamin_e'] ?? 0).toDouble(),
      vitaminK: (json['vitamin_k'] ?? 0).toDouble(),
      vitaminB1: (json['vitamin_b1'] ?? 0).toDouble(),
      vitaminB2: (json['vitamin_b2'] ?? 0).toDouble(),
      vitaminB6: (json['vitamin_b6'] ?? 0).toDouble(),
      vitaminB12: (json['vitamin_b12'] ?? 0).toDouble(),
      niacin: (json['niacin'] ?? 0).toDouble(),
      pantothenicAcid: (json['pantothenic_acid'] ?? 0).toDouble(),
      biotin: (json['biotin'] ?? 0).toDouble(),
      folate: (json['folate'] ?? 0).toDouble(),
      vitaminC: (json['vitamin_c'] ?? 0).toDouble(),
      recipeDetails: json['recipe_details'] != null
          ? RecipeDetails.fromJson(json['recipe_details'])
          : null,
    );
  }

  String get categoryDisplay {
    switch (category) {
      case '主食':
        return '主食';
      case '主菜':
        return '主菜';
      case '副菜':
        return '副菜';
      case '汁物':
        return '汁物';
      case 'デザート':
        return 'デザート';
      default:
        return category;
    }
  }
}

/// 料理と分量
class DishPortion {
  final Dish dish;
  final double servings;

  DishPortion({
    required this.dish,
    this.servings = 1.0,
  });

  factory DishPortion.fromJson(Map<String, dynamic> json) {
    return DishPortion(
      dish: Dish.fromJson(json['dish']),
      servings: (json['servings'] ?? 1.0).toDouble(),
    );
  }

  double get calories => dish.calories * servings;
  double get protein => dish.protein * servings;
  double get fat => dish.fat * servings;
  double get carbohydrate => dish.carbohydrate * servings;
}

File: backend/app/domain/entities/dish.py
"""Dish domain entities."""

from typing import Optional
from pydantic import BaseModel, Field

from app.domain.entities.enums import (
    MealTypeEnum,
    DishCategoryEnum,
    CookingMethodEnum,
)


class RecipeDetails(BaseModel):
    """レシピ詳細"""
    prep_time: Optional[int] = Field(default=None, description="下準備時間（分）")
    cook_time: Optional[int] = Field(default=None, description="調理時間（分）")
    servings: Optional[int] = Field(default=None, description="基準人数")
    steps: list[str] = Field(default_factory=list, description="調理手順")
    tips: Optional[str] = Field(default=None, description="コツ・ポイント")
    variations: Optional[str] = Field(default=None, description="アレンジ例")


class DishIngredient(BaseModel):
    """料理の材料"""
    food_id: int
    food_name: Optional[str] = None
    ingredient_id: Optional[int] = Field(default=None, description="基本食材ID（買い物リスト用）")
    ingredient_name: Optional[str] = Field(default=None, description="基本食材名（正規化された名前）")
    amount: float = Field(ge=0, description="g")
    display_amount: str = Field(default="", description="表示用の量（例: 1本, 1/2個）")
    unit: str = Field(default="g", description="単位（個, 本, 束, 枚, g）")
    cooking_method: CookingMethodEnum = CookingMethodEnum.RAW


class DishBase(BaseModel):
    """料理ベース"""
    name: str
    category: DishCategoryEnum
    meal_types: list[MealTypeEnum]
    serving_size: float = Field(default=1.0, ge=0.1)
    description: Optional[str] = None
    instructions: Optional[str] = None


class Dish(DishBase):
    """料理データモデル（栄養素計算済み）"""
    id: int
    ingredients: list[DishIngredient]
    # 作り置き関連
    storage_days: int = Field(default=1, description="作り置き可能日数（0=当日のみ）")
    min_servings: int = Field(default=1, description="最小調理人前")
    max_servings: int = Field(default=4, description="最大調理人前")
    # 味付け系統（段階的決定フロー用）
    flavor_profile: str = Field(default="和風", description="味付け系統（和風/洋風/中華）")
    # 計算済み栄養素（1人前あたり）
    calories: float = 0
    protein: float = 0
    fat: float = 0
    carbohydrate: float = 0
    fiber: float = 0
    # ミネラル
    sodium: float = 0
    potassium: float = 0
    calcium: float = 0
    magnesium: float = 0
    iron: float = 0
    zinc: float = 0
    # ビタミン
    vitamin_a: float = 0
    vitamin_d: float = 0
    vitamin_e: float = 0
    vitamin_k: float = 0
    vitamin_b1: float = 0
    vitamin_b2: float = 0
    vitamin_b6: float = 0
    vitamin_b12: float = 0
    niacin: float = 0
    pantothenic_acid: float = 0
    biotin: float = 0
    folate: float = 0
    vitamin_c: float = 0
    # レシピ詳細
    recipe_details: Optional[RecipeDetails] = Field(default=None, description="詳細レシピ")

    class Config:
        from_attributes = True

    def get_nutrient(self, nutrient: str) -> float:
        """栄養素の値を取得"""
        return getattr(self, nutrient, 0)


class DishPortion(BaseModel):
    """料理と分量"""
    dish: Dish
    servings: float = Field(default=1.0, ge=0.1, description="人前")

    def get_nutrient_total(self, nutrient: str) -> float:
        """栄養素の合計を取得"""
        return self.dish.get_nutrient(nutrient) * self.servings


class CookingFactor(BaseModel):
    """調理係数"""
    food_category: str
    cooking_method: CookingMethodEnum
    nutrient: str
    factor: float = Field(default=1.0, ge=0, le=2.0)
